File: firebase/functions/lib/community/resolve_join_request.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import '../utils/email_templates.dart';
import '../utils/firestore_utils.dart';
import '../utils/send_email_client.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/community/community.dart';
import 'package:data_models/community/membership.dart';
import 'package:data_models/community/membership_request.dart';
import 'package:data_models/utils/utils.dart';

class ResolveJoinRequest extends OnCallMethod<ResolveJoinRequestRequest> {
  ResolveJoinRequest()
      : super(
          'resolveJoinRequest',
          (jsonMap) => ResolveJoinRequestRequest.fromJson(jsonMap),
        );

  @override
  Future<void> action(
    ResolveJoinRequestRequest request,
    CallableContext context,
  ) async {
    return firestore.runTransaction((transaction) async {
      // get docs and verify admin permitted to approve / deny

      final modMembership = await firestoreUtils.getFirestoreObject(
        transaction: transaction,
        path:
            'memberships/${context.authUid}/community-membership/${request.communityId}',
        constructor: (map) => Membership.fromJson(map),
      );

      final userMembershipRequestSnapshot = await transaction.get(
        firestore.document(
          'community/${request.communityId}/join-requests/${request.userId}',
        ),
      );
      final userMembershipRequest = MembershipRequest.fromJson(
        userMembershipRequestSnapshot.data.toMap() ?? {},
      );
      final hasActiveRequest =
          userMembershipRequest.status == MembershipRequestStatus.requested;

      final userMembershipSnapshot = await transaction.get(
        firestore.document(
          'memberships/${request.userId}/community-membership/${request.communityId}',
        ),
      );
      final userMembership = userMembershipSnapshot.exists
          ? Membership.fromJson(
              firestoreUtils
                  .fromFirestoreJson(userMembershipSnapshot.data.toMap() ?? {}),
            )
          : Membership(
              userId: request.userId,
              communityId: request.communityId,
              status: MembershipStatus.nonmember,
              firstJoined: DateTime.now(),
            );

      if (!modMembership.isMod ||
          !hasActiveRequest ||
          userMembership.status?.isMember == true) {
        throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
      }

      // perform approval / denial

      if (request.approve == true) {
        final email = await constructApprovalEmail(
          transaction: transaction,
          userId: request.userId,
          communityId: request.communityId,
        );

        final userMembershipUpdated =
            userMembership.copyWith(status: MembershipStatus.member);
        if (userMembershipSnapshot.exists) {
          transaction.update(
            userMembershipSnapshot.reference,
            UpdateData.fromMap(
              jsonSubset(
                ['status'],
                firestoreUtils.toFirestoreJson(userMembershipUpdated.toJson()),
              ),
            ),
          );
        } else {
          transaction.set(
            userMembershipSnapshot.reference,
            DocumentData.fromMap(
              firestoreUtils.toFirestoreJson(userMembershipUpdated.toJson()),
            ),
          );
        }

        final userMembershipRequestUpdated = userMembershipRequest.copyWith(
          status: MembershipRequestStatus.approved,
        );
        if (userMembershipRequestSnapshot.exists) {
          transaction.update(
            userMembershipRequestSnapshot.reference,
            UpdateData.fromMap(
              jsonSubset(
                ['status'],
                firestoreUtils.toFirestoreJson(
                  userMembershipRequestUpdated.toJson(),
                ),
              ),
            ),
          );
        } else {
          transaction.set(
            userMembershipRequestSnapshot.reference,
            DocumentData.fromMap(
              firestoreUtils
                  .toFirestoreJson(userMembershipRequestUpdated.toJson()),
            ),
          );
        }

        await sendEmailClient.sendEmail(email, transaction: transaction);
      } else {
        final userMembershipRequestUpdated = userMembershipRequest.copyWith(
          status: MembershipRequestStatus.denied,
        );
        if (userMembershipRequestSnapshot.exists) {
          transaction.update(
            userMembershipRequestSnapshot.reference,
            UpdateData.fromMap(
              jsonSubset(
                ['status'],
                firestoreUtils.toFirestoreJson(
                  userMembershipRequestUpdated.toJson(),
                ),
              ),
            ),
          );
        } else {
          transaction.set(
            userMembershipRequestSnapshot.reference,
            DocumentData.fromMap(
              firestoreUtils
                  .toFirestoreJson(userMembershipRequestUpdated.toJson()),
            ),
          );
        }
      }
    });
  }

  Future<SendGridEmail> constructApprovalEmail({
    required Transaction transaction,
    required String userId,
    required String communityId,
  }) async {
    final users = await firestoreUtils.getUsers([userId]);
    final community = await firestoreUtils.getFirestoreObject(
      path: 'community/$communityId',
      constructor: (map) => Community.fromJson(map),
    );

    final emailAddress = users.single.email;
    final message = SendGridEmailMessage(
      subject: 'Join Request Approved: ${community.name ?? community.id}',
      html: makeJoinApprovedBody(community: community),
    );
    final noReplyEmailAddr =
        functions.config.get('app.no_reply_email') as String;
    return SendGridEmail(
      to: [emailAddress],
      from: '${community.name ?? community.id} <$noReplyEmailAddr>',
      message: message,
    );
  }
}

File: firebase/functions/src/community/resolve_join_request.ts
import * as functions from 'firebase-functions';
import { HttpsError } from 'firebase-functions/lib/providers/https';
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils';
import { sendEmailClient } from '../utils/send_email_client';
import { firebaseAuthUtils } from '../utils/infra/firebase_auth_utils';
import { OnCallMethod } from '../on_call_function';
import {
  ResolveJoinRequestRequest, Membership, MembershipStatus, MembershipRequest,
  MembershipRequestStatus, Community,
} from '../types';
import { membershipIsMod } from '../types';
import { jsonSubset } from '../utils/utils';

export class ResolveJoinRequest extends OnCallMethod<ResolveJoinRequestRequest> {
  constructor() {
    super('resolveJoinRequest', (json) => json as ResolveJoinRequestRequest);
  }

  async action(
    request: ResolveJoinRequestRequest,
    context: functions.https.CallableContext
  ): Promise<void> {
    await firestore.runTransaction(async (transaction) => {
      // Get mod membership
      const modMembershipDoc = await transaction.get(
        firestore.doc(`memberships/${context.auth?.uid}/community-membership/${request.communityId}`)
      );
      const modMembership = modMembershipDoc.data() as Membership | undefined;

      // Get join request
      const userMembershipRequestRef = firestore.doc(
        `community/${request.communityId}/join-requests/${request.userId}`
      );
      const userMembershipRequestSnap = await transaction.get(userMembershipRequestRef);
      const userMembershipRequest = userMembershipRequestSnap.data() as MembershipRequest | undefined;
      const hasActiveRequest = userMembershipRequest?.status === MembershipRequestStatus.requested;

      // Get user membership
      const userMembershipRef = firestore.doc(
        `memberships/${request.userId}/community-membership/${request.communityId}`
      );
      const userMembershipSnap = await transaction.get(userMembershipRef);
      const userMembership: Membership = userMembershipSnap.exists
        ? (userMembershipSnap.data() as Membership)
        : { userId: request.userId, communityId: request.communityId, status: MembershipStatus.nonmember, firstJoined: new Date() };

      const isMod = modMembership?.status ? membershipIsMod(modMembership.status) : false;
      if (!isMod || !hasActiveRequest || userMembership.status === MembershipStatus.member) {
        throw new HttpsError('failed-precondition', 'unauthorized');
      }

      if (request.approve) {
        // Update user membership to member
        const updatedMembership = { ...userMembership, status: MembershipStatus.member };
        if (userMembershipSnap.exists) {
          transaction.update(
            userMembershipRef,
            jsonSubset(['status'], firestoreUtils.toFirestoreJson(updatedMembership as unknown as Record<string, unknown>))
          );
        } else {
          transaction.set(
            userMembershipRef,
            firestoreUtils.toFirestoreJson(updatedMembership as unknown as Record<string, unknown>)
          );
        }

        // Update request to approved
        const updatedRequest = { ...userMembershipRequest, status: MembershipRequestStatus.approved };
        if (userMembershipRequestSnap.exists) {
          transaction.update(
            userMembershipRequestRef,
            jsonSubset(['status'], firestoreUtils.toFirestoreJson(updatedRequest as unknown as Record<string, unknown>))
          );
        } else {
          transaction.set(
            userMembershipRequestRef,
            firestoreUtils.toFirestoreJson(updatedRequest as unknown as Record<string, unknown>)
          );
        }
      } else {
        // Deny
        const updatedRequest = { ...userMembershipRequest, status: MembershipRequestStatus.denied };
        if (userMembershipRequestSnap.exists) {
          transaction.update(
            userMembershipRequestRef,
            jsonSubset(['status'], firestoreUtils.toFirestoreJson(updatedRequest as unknown as Record<string, unknown>))
          );
        } else {
          transaction.set(
            userMembershipRequestRef,
            firestoreUtils.toFirestoreJson(updatedRequest as unknown as Record<string, unknown>)
          );
        }
      }
    });
  }
}
